packages/core/src/validate/run.ts

```diff
@@ -557,3 +560,32 @@ function* getLinkingErrors(
     }
   }
 }
+
+function* getInternalFunctionStorageErrors(
+  contractDef: ContractDefinition,
+  decodeSrc: SrcDecoder,
+): Generator<ValidationError> {
+  // Note: Solidity does not allow annotations for non-public state variables, so this cannot be skipped with annotations
+  for (const variableDef of findAll('VariableDeclaration', contractDef)) {
+    if (variableDef.typeName?.nodeType === 'FunctionTypeName' && variableDef.typeName.visibility === 'internal') {
+      yield {
+        kind: 'internal-function-storage',
+        name: variableDef.name,
+        src: decodeSrc(variableDef),
+      };
+    }
+  }
+
+  // Note: Solidity currently does not include struct annotations in the AST, so this cannot be skipped with annotations
+  for (const structDef of findAll('StructDefinition', contractDef)) {
```

Comment:
Pretty sure this loop is redundant. `StructDefinition.members` is `VariableDeclaration[]` so each member should be covered by the `findAll` loop above.

Reply:
This makes me realize the `findAll` usage above would miss structs that are declared outside of the contract.

We may need to look for all storage variables which use structs, and recursively check their referenced structs for function types.

I think we should still continue to look for struct declarations within contracts though, because for example namespaced storage structs could define these and not be referenced by regular storage variables.

Reply:
> We may need to look for all storage variables which use structs, and recursively check their referenced structs for function types.

Yes this sounds right.